src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -1604,12 +1616,40 @@ def generate(
         ):
             forced_decoder_ids = self.generation_config.forced_decoder_ids
 
-        if generation_config.return_timestamps:
-            logits_processor = [WhisperTimeStampLogitsProcessor(generation_config)]
-
         if len(forced_decoder_ids) > 0:
             generation_config.forced_decoder_ids = forced_decoder_ids
 
+        if prompt_ids is not None:
+            if kwargs.get("decoder_start_token_id") is not None:
+                raise ValueError(
+                    "When specifying `prompt_ids`, you cannot also specify `decoder_start_token_id` as it gets overwritten."
+                )
+            prompt_ids = to_py_obj(prompt_ids)
+            decoder_start_token_id, *text_prompt_ids = prompt_ids
+            # Set the decoder_start_token_id to <|startofprev|>
+            kwargs.update({"decoder_start_token_id": decoder_start_token_id})
+
+            # Update the max generation length to include the prompt
+            specified_max_length = kwargs.pop("max_new_tokens", None) or kwargs.pop("max_length", None)
+            default_max_length = generation_config.max_new_tokens or generation_config.max_length
+            non_prompt_max_length = specified_max_length or default_max_length
+            generation_config.max_new_tokens = non_prompt_max_length + len(text_prompt_ids)
+
+            # Reformat the forced_decoder_ids to incorporate the prompt
+            non_prompt_forced_decoder_ids = (
+                kwargs.pop("forced_decoder_ids", None) or generation_config.forced_decoder_ids
+            )
```

Comment:
Nice! I think this now supports the different ways that `forced_decoder_ids` may be passed in?

1. Through `model.config.forced_decoder_ids = processor.get_decoder_prompt_ids(language=..., task=...)`

2. Through `model.generate(input_features, forced_decoder_ids=forced_decoder_ids)`

3. Through `model.generate(input_features, language=..., task=...)`

It would be good if there are unit tests for these different methods.

Reply:
I don't believe `model.generate` allows passing in task or language directly as in 3. above, but I've added tests for the other two

Reply:
It does allow that (and I think it might even be the preferred method now) but for some reason the language needs to be the token, such as `"<|ja|>"` rather than `"ja"`.

Reply:
Note that the language code change was @connor-henderson's most recent PR! This `forced_generation_ids` logic is in-place so that the code is backward compatible with our previous way of handling the langauge/task, where we either set it in the config as `config.forced_decoder_ids`, or explicitly as `forced_decoder_ids` to the generate method (see https://github.com/huggingface/transformers/pull/21965#discussion_r1127907439)

Reply:
haha derp, I didn't look at the author 🙈 my bad!